File: 2020/ruby/08.rb
require './utils.rb'
l=read(8).split "\n"
it = 0
acc = 0
seen = Set.new
while !seen.include?(it)
  seen.add it
  a, b = l[it].split
  case a
  when 'jmp' then it += b.to_i;
  when 'acc' then acc += b.to_i; it += 1;
  when 'nop' then it += 1;
  end
end
puts acc

(0...l.size).each{|i|
  it = 0
  acc = 0
  seen = Set.new
  while !seen.include?(it) && 0 <= it && it < l.size
    seen.add it
    a, b = l[it].split
    if i == it && a != 'acc'
      a = a == 'jmp' ? 'nop' : 'jmp'
    end
    case a
    when 'jmp' then it += b.to_i;
    when 'acc' then acc += b.to_i; it += 1;
    when 'nop' then it += 1;
    end
  end
  puts acc if !seen.include? it
}

File: 2020/python/08.py
from utils import *
from aocd import data

lines = data.split("\n")

seen = set()
it = 0
acc = 0
while it not in seen:
    seen.add(it)
    a, b = lines[it].split()
    if a == "acc":
        acc += int(b)
    elif a == "jmp":
        it += int(b) - 1
    it += 1
print(acc)

for i in range(len(lines)):
    seen = set()
    it = 0
    acc = 0
    while 0 <= it < len(lines) and it not in seen:
        seen.add(it)
        a, b = lines[it].split()
        if it == i:
            a = "jmp" if a == "nop" else "nop"
        if a == "acc":
            acc += int(b)
        elif a == "jmp":
            it += int(b) - 1
        it += 1
    if not (0 <= it < len(lines)):
        print(acc)
